### PR TITLE
initial working role

### PR DESCRIPTION
**Pull request by fujexo**

### CODE REVIEW

File: defaults/main.yml [hunk 1]
```diff
@@ -1 +1,5 @@
 ---
+
+icinga2_client_api_url: "http://monitoring.example.com:5665"
```
Comment (karras) [inline defaults/main.yml]: Add variable descriptions...

File: handlers/main.yml [hunk 2]
```diff
@@ -1 +1,6 @@
 ---
+
+- name: reload icinga2 daemon
```
Comment (karras) [inline handlers/main.yml]: Adjust the handler name to our standard

File: README.rst [hunk 3]
```diff
@@ -1,15 +1,15 @@
...
-  :target: https://travis-ci.org/adfinis-sygroup/ansible-role-_template
+.. image:: https://img.shields.io/travis/adfinis-sygroup/ansible-role-icinga2_client.svg?style=flat-square
+  :target: https://travis-ci.org/adfinis-sygroup/ansible-role-icinga2_client
 
-.. image:: https://img.shields.io/badge/galaxy-adfinis--sygroup._template-660198.svg?style=flat-square
-  :target: https://galaxy.ansible.com/adfinis-sygroup/_template
+.. image:: https://img.shields.io/badge/galaxy-adfinis--sygroup.icinga2_client-660198.svg?style=flat-square
+  :target: https://galaxy.ansible.com/adfinis-sygroup/icinga2_client
 
 A brief description of the role goes here.
```
Comment (karras) [inline README.rst]: Describe what the role does and which dependencies it has...

File: meta/main.yml [hunk 4]
```diff
@@ -23,4 +24,5 @@ galaxy_info:
       versions:
         - 6
         - 7
-  galaxy_tags: []
+  galaxy_tags:
+    - "icinga2"
```
Comment (karras) [inline meta/main.yml]: Add a few more tags ;)

File: tasks/certificates.yml [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    body: '{ "cn": "{{ inventory_hostname }}"}'
+    force_basic_auth: yes
+    status_code: 200
+    body_format: 'json'
+    headers:
+      Accept: 'application/json'
+    validate_certs : no
+  register: icinga2_client_ticket
+
+# - name: generate ticket on master
```
Comment (karras) [inline tasks/certificates.yml]: Remove those leftovers...

File: tasks/certificates.yml [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+#   shell: "icinga2 pki ticket --cn {{ inventory_hostname }} --salt {{ icinga2_salt }}"
+#   register: icinga2_client_ticket
+#   delegate_to: "{{ icinga2_master_fqdn }}"
+
+- name: Ticket generated!
```
Comment (karras) [inline tasks/certificates.yml]: Task names should be all lowercase according to our standard...

File: tasks/certificates.yml [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+  debug:
+    msg: "{{ icinga2_client_ticket }}"
+
+- name: Ticket and fqdn
+  debug:
+    msg: "{{ icinga2_client_ticket }} : is for {{ inventory_hostname }}"
+
+# - name: create pki folder on clients
```
Comment (karras) [inline tasks/certificates.yml]: Leftovers should be eaten...

File: tasks/configuration.yml [hunk 8]
```diff
@@ -1 +1,25 @@
 ---
+
+- name: remove conf.d folder
+  file:
+    path: "/etc/icinga2/conf.d"
```
Comment (karras) [inline tasks/configuration.yml]: Don't use ticket unless required

File: tasks/configuration.yml [hunk 9]
```diff
@@ -1 +1,25 @@
...
+    state: absent
+  notify: "reload icinga2 daemon"
+
+- name: install icinga2 configuration
+  template:
+    src: templates/etc/icinga2/icinga2.conf.j2
+    dest: /etc/icinga2/icinga2.conf
```
Comment (karras) [inline tasks/configuration.yml]: SELinux context + label are missing, also define the file permissions

File: tasks/configuration.yml [hunk 10]
```diff
@@ -1 +1,25 @@
...
+  notify: "reload icinga2 daemon"
+
+- name: install icinga2 zones configuration
+  template:
+    src: templates/etc/icinga2/zones.conf.j2
+    dest: /etc/icinga2/zones.conf
```
Comment (karras) [inline tasks/configuration.yml]: SELinux context + label are missing, also define the file permissions

File: tasks/configuration.yml [hunk 11]
```diff
@@ -1 +1,25 @@
...
+  notify: "reload icinga2 daemon"
+
+- name: install icinga2 constants configuration
+  template:
+    src: templates/etc/icinga2/constants.conf.j2
```
Comment (karras) [inline tasks/configuration.yml]: SELinux context + label are missing, also define the file permissions

File: tasks/main.yml [hunk 12]
```diff
@@ -6,16 +6,21 @@
...
+    - 'role::icinga2_client:install'
 
 - import_tasks: configuration.yml
   tags:
-    - 'role::$role'
-    - 'role::$role:config'
+    - 'role::icinga2_client'
+    - 'role::icinga2_client:config'
+
+- import_tasks: certificates.yml
```
Comment (karras) [inline tasks/main.yml]: Is this idempotent? I think you only need to get a new ticket if the agent packages have been installed and if there's no certificate in place yet?

File: templates/etc/icinga2/constants.conf.j2 [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+/* Our local instance name. By default this is the server's hostname as returned by `hostname --fqdn`.
+ * This should be the common name from the API certificate.
+ */
+const NodeName = "{{ inventory_hostname }}"
+
+/* Our local zone name. */
+const ZoneName = "{{ inventory_hostname }}"
+
+/* Secret key for remote node tickets */
+const TicketSalt = "{{ icinga2_salt|default("") }}"
```
Comment (karras) [inline templates/etc/icinga2/constants.conf.j2]: Undefined variable?

File: templates/etc/icinga2/zones.conf.j2 [hunk 14]
```diff
@@ -0,0 +1,17 @@
+/**
+ *  {{ ansible_managed }}
+ */
+// todo satellite zone
+
+object Zone "global-templates" {
+  global = true
+}
+
+object Endpoint "icinga-master" {
```
Comment (karras) [inline templates/etc/icinga2/zones.conf.j2]: This needs to be customizable right?

File: handlers/main.yml (hunk 2 above)
Comment (fujexo) [reply handlers/main.yml]: Our ansible_guide is a bit unclear about this. Did you mean to rename the handler to `$role_name reload icinga2 daemon`?

File: tasks/certificates.yml (hunk 5 above)
Comment (fujexo) [reply tasks/certificates.yml]: is not yet finished and will be update once the role is working properly

File: tasks/main.yml (hunk 12 above)
Comment (fujexo) [reply tasks/main.yml]: This is not yet finished and will be updated once the role is working properly

File: templates/etc/icinga2/constants.conf.j2 (hunk 13 above)
Comment (fujexo) [reply templates/etc/icinga2/constants.conf.j2]: I'm not yet sure if the TicketSalt must be configured on clients, so the default is empty. I will check and update this if needed after more testing

**Comment by fujexo** [comment]
Thanks @karras for your input. I've updated most of the files and will request additional feedback once I've got a good initial working role :)

**Comment by fujexo** [comment]
@karras @MrStrix 
The role is now functional. I would appreciate your review!

File: README.rst [hunk 15]
```diff
@@ -1,68 +1,73 @@
...
+  # The user needs at least permissions to create ticket tokens.
+  icinga2_client_api_user: "root"
 
-A description of the settable variables for this role should go here, including
-any variables that are in defaults/main.yml, vars/main.yml, and any variables
-that can/should be set via parameters to the role. Any variables that are read
-from other roles and/or the global scope (ie. hostvars, group vars, etc.)
-should be mentioned here as well.
+  # The API password of the icinga2 master
+  icinga2_client_api_pass: "v3rys3cur3p455"
```
Comment (karras) [inline README.rst]: Please replace it with a more obvious example password ;)

File: meta/main.yml [hunk 16]
```diff
@@ -23,4 +24,11 @@ galaxy_info:
...
+    - "client"
+    - "monitoring"
+    - "certificates"
+    - "adfinis"
+    - "sygroup"
+    - "adfinis-sygroup.ch"
```
Comment (karras) [inline meta/main.yml]: I'd suggest to remove company related tags.

File: tasks/certificates.yml [hunk 17]
```diff
@@ -0,0 +1,41 @@
+---
+
+# TODO: Only request ticket when needed
```
Comment (karras) [inline tasks/certificates.yml]: How about checking if a certain file (e.g. cert) exists and register a variable?

File: tasks/configuration.yml [hunk 18]
```diff
@@ -1 +1,47 @@
 ---
+
+# Icinga2 currently supports only two endpoints.
+# https://github.com/Icinga/icinga2/issues/3533
+- name: fail when more than 2 icinga2 endpoints are defined
+  fail:
+    msg: No more than 2 Endpoints allowed
```
Comment (karras) [inline tasks/configuration.yml]: s/Endpoints/endpoints

File: tasks/main.yml [hunk 19]
```diff
@@ -6,16 +6,16 @@
     - '{{ ansible_distribution }}_{{ ansible_distribution_major_version }}.yml'
     - '{{ ansible_os_family }}.yml'
   tags:
-    - 'role::$role'
-    - 'role::$role:install'
-    - 'role::$role:config'
+    - 'role::icinga2_client'
+    - 'role::icinga2_client:install'
```
Comment (karras) [inline tasks/main.yml]: Remove the `install` tag

File: templates/etc/icinga2/constants.conf.j2 [hunk 20]
```diff
@@ -0,0 +1,32 @@
+/**
+ *  {{ ansible_managed }}
```
Comment (karras) [inline templates/etc/icinga2/constants.conf.j2]: Should we add a comment from which icinga2 package version the initial config file was taken? Thus it's easier to compare them later on in operations.

**Comment by karras** [comment]
Just some more nitpicking, other than that it looks nice!

File: templates/etc/icinga2/constants.conf.j2 (hunk 20 above)
Comment (fujexo) [reply templates/etc/icinga2/constants.conf.j2]: Is probably a good idea

File: tasks/certificates.yml [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+    user: '{{ icinga2_client_api_user }}'
+    password: '{{ icinga2_client_api_pass }}'
+    body: '{ "cn": "{{ inventory_hostname }}"}'
+    force_basic_auth: yes
+    status_code: 200
+    body_format: 'json'
+    headers:
+      Accept: 'application/json'
+    validate_certs : no
+  register: icinga2_client_register_ticket
```
Comment (karras) [inline tasks/certificates.yml]: This var is no longer required right?

**Pull request merged**